File: creational/factory_method/python/payment_processors.py
from abc import ABC, abstractmethod
import re

class PaymentProcessor(ABC):
    @abstractmethod
    def validate_data(self, data: dict[str, str]) -> bool:
        pass

    @abstractmethod
    def calculate_commission(self, amount: float) -> float:
        pass
    
    @abstractmethod
    def process_payment(self, data: dict[str, str | float]) -> dict[str, str | bool | float]:
        pass

    @abstractmethod
    def generate_receipt(self, data: dict[str, str | float]) -> str:
        pass



class CreditCardProcessor(PaymentProcessor):
    def validate_data(self, data: dict[str, str]) -> bool:
        required_fields = ["card_number", "expiry_date", "cvv"]
        is_valid = all(field in data for field in required_fields)
        cvv_is_valid = len(data.get("cvv")) >= 3 # type: ignore
        return is_valid and cvv_is_valid

    def calculate_commission(self, amount: float) -> float:
        return amount * 0.03  # 3% commission
    
    def process_payment(self, data: dict[str, str | float]) -> dict[str, str | bool | float]:
        return {
            "success": True,
            "transaction_id": "CC-" + str(hash(str(data))),
            "amount": data.get("amount", 0.0)
        }

    def generate_receipt(self, data: dict[str, str | float]) -> str:
        return f"Credit Card Payment Receipt\nAmount: {"%.2f" % round(float(data.get('amount', 0.00)), 2)} \nTransaction ID: {data.get('transaction_id', 'N/A')}"

class PayPalProcessor(PaymentProcessor):
    def validate_data(self, data: dict[str, str]) -> bool:
        required_fields = ["email", "password"]
        return all(field in data for field in required_fields)

    def calculate_commission(self, amount: float) -> float:
        return amount * 0.02  # 2% commission
    
    def process_payment(self, data: dict[str, str | float]) -> dict[str, str | bool | float]:
        return {
            "success": True,
            "transaction_id": "PP-" + str(hash(str(data))),
            "amount": data.get("amount", 0.0)
        }

    def generate_receipt(self, data: dict[str, str | float]) -> str:
        return f"PayPal Payment Receipt\nAmount: {"%.2f" % round(float(data.get('amount', 0.00)), 2)}\nTransaction ID: {data.get('transaction_id', 'N/A')}"

class BankTransferProcessor(PaymentProcessor):
    def validate_data(self, data: dict[str, str]) -> bool:
        required_fields = ["account_number", "routing_number", "account_holder"]
        present_data = all(field in data for field in required_fields)
        account_number_is_valid = len(data.get("account_number", "")) > 9
        return present_data and account_number_is_valid
    def calculate_commission(self, amount: float) -> float:
        return amount * 0.01  # 1% commission
    
    def process_payment(self, data: dict[str, str | float]) -> dict[str, str | bool | float]:
        return {
            "success": True,
            "transaction_id": "BT-" + str(hash(str(data))),
            "amount": data.get("amount", 0.0)
        }

    def generate_receipt(self, data: dict[str, str | float]) -> str:
        return f"Bank Transfer Receipt\nAmount: {"%.2f" % round(float(data.get('amount', 0.00)), 2)}\nTransaction ID: {data.get('transaction_id', 'N/A')}"

class DigitalWalletProcessor(PaymentProcessor):
    def _validate_phone_number(self, phone: str) -> bool:
        """
        Validate phone number format.
        Supports formats:
        - International: +1-234-567-8900, +1.234.567.8900, +1 234 567 8900
        - National: (234) 567-8900, 234-567-8900, 234.567.8900
        - Simple: 2345678900
        """
        # Remove any spaces, dashes, dots, or parentheses
        phone = re.sub(r'[\s\-\.\(\)]', '', phone)
        
        # Check if it's an international number (starts with +)
        if phone.startswith('+'):
            # International format: +1-234-567-8900 (10-15 digits after +)
            pattern = r'^\+\d{10,15}$'
        else:
            # National format: (234) 567-8900 (10 digits)
            pattern = r'^\d{10}$'
        
        return bool(re.match(pattern, phone))

    def validate_data(self, data: dict[str, str]) -> bool:
        required_fields = ["wallet_id", "phone_number"]
        if not all(field in data for field in required_fields):
            return False
        
        # Validate phone number format
        return self._validate_phone_number(data["phone_number"])

    def calculate_commission(self, amount: float) -> float:
        return amount * 0.015  # 1.5% commission
    
    def process_payment(self, data: dict[str, str | float]) -> dict[str, str | bool | float]:
        return {
            "success": True,
            "transaction_id": "DW-" + str(hash(str(data))),
            "amount": data.get("amount", 0.0)
        }

    def generate_receipt(self, data: dict[str, str | float]) -> str:
        return f"Digital Wallet Payment Receipt\nAmount: {"%.2f" % round(float(data.get('amount', 0.00)), 2)}\nTransaction ID: {data.get('transaction_id', 'N/A')}"

File: creational/factory_method/csharp/src/PaymentProcessors.cs
public class CreditCardProcessor : IPaymentProcessor<CreditCardData>
{
    public bool ValidateData(CreditCardData paymentData)
    {
        return !string.IsNullOrEmpty(paymentData.CardNumber) &&
                   paymentData.CardNumber.Length == 16 &&
                   !string.IsNullOrEmpty(paymentData.ExpiryDate) &&
                   paymentData.Cvv != null && paymentData.Cvv.Length == 3;
    }
    public double CalculateCommission(double amount)
    {
        return amount * 0.03;
    }
    public PaymentResult ProcessPayment(CreditCardData paymentData)
    {
        if (ValidateData(paymentData))
        {
            string transactionId = Guid.NewGuid().ToString();
            return new PaymentResult { Success = true, TransactionId = transactionId, Message = "Credit card payment successfully processed." };
        }
        return new PaymentResult { Success = false, TransactionId = "", Message = "Invalid credit card information." };
    }
    public string GenerateReceipt(CreditCardData paymentData)
    {
        return $"Credit Card Payment:\n" +
                $"Amount: {paymentData.Amount:F2}\n" +
                $"Lasts 4 digits of your card: {paymentData.CardNumber?.Substring(paymentData.CardNumber.Length - 4)}\n" +
                $"Transaction ID: {paymentData.TransactionId}";
    }
}

public class PayPalProcessor : IPaymentProcessor<PayPalData>
{
    public bool ValidateData(PayPalData paymentData)
    {
        return !string.IsNullOrEmpty(paymentData.Email) &&
                    paymentData.Email.Contains("@");
    }
    public double CalculateCommission(double amount)
    {
        return amount * 0.02;
    }
    public PaymentResult ProcessPayment(PayPalData paymentData)
    {
        if (ValidateData(paymentData))
        {
            string transactionId = Guid.NewGuid().ToString();
            return new PaymentResult { Success = true, TransactionId = transactionId, Message = "PayPal payment successfully processed." };
        }
        return new PaymentResult { Success = false, TransactionId = "", Message = "Invalid PayPal data." };
    }
    public string GenerateReceipt(PayPalData paymentData)
    {
        return $"PayPal Payment:\n" +
                $"Amount: {paymentData.Amount:F2}\n" +
                $"Email: {paymentData.Email}\n" +
                $"Transaction ID: {paymentData.TransactionId}";
    }
}

public class BankTransferProcessor : IPaymentProcessor<BankTransferData>
{
    public bool ValidateData(BankTransferData paymentData)
    {
        return !string.IsNullOrEmpty(paymentData.AccountNumber) &&
                   paymentData.AccountNumber.Length >= 10;
    }
    public double CalculateCommission(double amount)
    {
        return amount * 0.01;
    }
    public PaymentResult ProcessPayment(BankTransferData paymentData)
    {
        if (ValidateData(paymentData))
        {
            string transactionId = Guid.NewGuid().ToString();
            return new PaymentResult { Success = true, TransactionId = transactionId, Message = "Bank transfer payment successfully processed." };
        }
        return new PaymentResult { Success = false, TransactionId = "", Message = "Invalid bank transfer data." };
    }
    public string GenerateReceipt(BankTransferData paymentData)
    {
        return $"Bank Transfer:\n" +
                $"Amount: {paymentData.Amount:F2}\n" +
                $"Account number: {paymentData.AccountNumber}\n" +
                $"Transaction ID: {paymentData.TransactionId}";
    }
}

public class DigitalWalletProcessor : IPaymentProcessor<DigitalWalletData>
{
    public bool ValidateData(DigitalWalletData paymentData)
    {
        return !string.IsNullOrEmpty(paymentData.WalletId) &&
                   !string.IsNullOrEmpty(paymentData.PhoneNumber) &&
                   paymentData.PhoneNumber.StartsWith("+");
    }
    public double CalculateCommission(double amount)
    {
        return amount * 0.015;
    }
    public PaymentResult ProcessPayment(DigitalWalletData paymentData)
    {
        if (ValidateData(paymentData))
        {
            string transactionId = Guid.NewGuid().ToString();
            return new PaymentResult { Success = true, TransactionId = transactionId, Message = "Digital wallet payment successfully processed." };
        }
        return new PaymentResult { Success = false, TransactionId = "", Message = "Invalid digital wallet data." };
    }
    public string GenerateReceipt(DigitalWalletData paymentData)
    {
        return $"Digital Wallet Payment:\n" +
                $"Amount: {paymentData.Amount:F2}\n" +
                $"Wallet ID: {paymentData.WalletId}\n" +
                $"Transaction ID: {paymentData.TransactionId}";
    }
}